ocaml/xapi/xapi_sr.ml

```diff
@@ -778,12 +778,20 @@ let scan ~__context ~sr =
               Db.VDI.get_records_where ~__context
                 ~expr:(Eq (Field "SR", Literal sr'))
             in
+            let string_of_l lst =
+              List.map (fun (_, v) -> v.vDI_uuid) lst |> String.concat ","
+            in
             (* It is sufficient to just compare the refs in two db_vdis, as this
                is what update_vdis uses to determine what to delete *)
             let vdis_ref_equal db_vdi1 db_vdi2 =
-              Listext.List.set_difference (List.map fst db_vdi1)
-                (List.map fst db_vdi2)
-              = []
+              let refs1 = List.map fst db_vdi1 in
+              let refs2 = List.map fst db_vdi2 in
+
+              (* VDIs that are in db_vdi1 but not in db_vdi2 *)
+              let vdis_removed = Listext.List.set_difference refs1 refs2 in
```

Comment:
Are these lists short? `set_difference` is O(n^2). When nothing is added and nothing is removed, why is this not just set equality, i.e., sorted lists are equal?

Reply:
There can be thousands of VDIs in a system, definitely lists should not be used here

Reply:
I like the idea of set equality/sorted lists are equal

Reply:
I modified the code to use Set instead.

Reply:
in fact I wanted to add (and I just pushed it now) some debug messages. Because as Pau said there can be hundreds of VDIs so it can helps to show the ones that have been added or removed. And probably I can remove the debug message that write the list before and after. What do you think?

Reply:
And yes set  equality looks good now. Because if during the scan a VDI has been added and removed sets are not empty but we don't need to start a new scan because the list of VDIs is ok. So yes I'm using equality now.